### PR TITLE
General maintenance for chem master

### PR DESCRIPTION
**Pull request by SyncIt21**
## About The Pull Request

**1. Qol**
- Adds screen tips & examines for screwdriver, wrench, crowbar & beaker insertion, removal & replacing actions
- Analyzing reagents is now a client side feature & not a back end mode, meaning one person can see details of a reagent while the other can print stuff and do other operations so it's a non blocking operation. This also means 2 players can see information of 2 different reagents in their own screens, With that the overlay for analysis mode has been removed
- You cannot do any tool acts on machines while printing. Balloon alerts will be displayed warning you of that.
- The preferred container for the master reagent in the beaker is now showed in both condiment & chem master. It can be enabled/disabled via a CheckBox

**2. Code Improvements**
- Removed defines like `TARGET_BEAKER` , `TARGET_BEAKER` etc. ther functionality is implemented as params in the `transfer_reagent()` proc directly
- Removed all variables relating to analyzing reagents like `reagent_analysis_mode`, `has_container_suggestion` etc. all memory savings
- `printable_containers` now stores static values that can be shared across many chem masters
- Updates only overlays and not the whole icon during operations for efficiency

**3. Fixes**
 - You can hit the chem master with the screwdriver, wrench, crowbar & beaker when in combat mode
 - You cannot insert hologram items into the chem master
 - Deconstructing a condiment master will give you the circuit board already pre-programmed with that option selected so you don't need to use a screwdriver to re program it
 - `printing_amount` is now the maximum number of containers that can be printed at a time. Presently this number with upgraded parts would print out empty containers especially for patches. This is because `volume_per_item` does not take into consideration this var. Also this var would not give control to the player on exactly how many containers to print as whatever amount the player entered would be multiplied with this value producing a lot of waste & worse empty containers. Now this var determines exactly how many containers you can print and is imposed on the client side UI as well

**4. Refactors (UI performance)**
- Beaker data is compressed into a single entity & sent to the UI. This is set to null if no beaker is loaded thus saving data sent
- Reuses Beaker props from chem synthesizer to reduce code
- reagent REF replaced with direct type converted to text and later converted with `text2path()` cause its much faster

## Changelog
:cl:
qol: Adds screen tips & examines for screwdriver, wrench, crowbar & beaker insertion, removal & replacing actions
qol: Analyzing reagents no longer blocks other players from doing other operations. Multiple players can analyze different reagents on the same machine
qol: You cannot do any tool acts on the machine while printing to prevent any side effects.
qol: The preferred container for the master reagent in the beaker is now showed in both condiment & chem master. The feature can be enabled/disabled via a check box
code: removed defines for reagent transfer, vars for reagent analyzis to save memory. Autodoc for other vars & procs
fix: You can hit the chem master with tools like screwdriver, crowbar, wrench & beaker in combat mode
fix: You cannot insert hologram items into the chem master
fix: Deconstructing a condiment master will give you the circuit board already pre-programmed with that option
fix: You now print the exact amount of containers requested even with upgraded parts without creating empty containers. Max printable containers is 13 with tier 4 parts able to print 50 containers.
refactor: Optimized client side UI code & chem master as a whole.
/:cl:

### CODE REVIEW

File: code/modules/reagents/chemistry/machinery/chem_master.dm [hunk 1]
```diff
@@ -1,84 +1,97 @@
...
+	if(!QDELETED(beaker))
+		beaker.forceMove(drop_location())
+
+/obj/machinery/chem_master/add_context(atom/source, list/context, obj/item/held_item, mob/user)
+	. = NONE
+	if(isnull(held_item) || (held_item.item_flags & ABSTRACT) || (held_item.flags_1 & HOLOGRAM_1))
+		if(isnull(held_item))
+			context[SCREENTIP_CONTEXT_RMB] = "Remove beaker"
+			. = CONTEXTUAL_SCREENTIP_SET
+		return
```
Comment (MrMelbert) [inline code/modules/reagents/chemistry/machinery/chem_master.dm:60]:
```suggestion
		return .
```

File: code/modules/reagents/chemistry/machinery/chem_master.dm [hunk 2]
```diff
@@ -114,39 +125,119 @@
...
+	return ..()
+
 /obj/machinery/chem_master/wrench_act(mob/living/user, obj/item/tool)
+	if(user.combat_mode)
+		return NONE
+
+	. = ITEM_INTERACT_BLOCKING
+	if(is_printing)
+		balloon_alert(user, "still printing")
+		return
```
Comment (MrMelbert) [inline code/modules/reagents/chemistry/machinery/chem_master.dm:195]:
```suggestion
		balloon_alert(user, "still printing!")
		return .
```

File: code/modules/reagents/chemistry/machinery/chem_master.dm [hunk 3]
```diff
@@ -114,39 +125,119 @@
...
-	return ITEM_INTERACT_BLOCKING
 
 /obj/machinery/chem_master/screwdriver_act(mob/living/user, obj/item/tool)
+	if(user.combat_mode)
+		return NONE
+
+	. = ITEM_INTERACT_BLOCKING
+	if(is_printing)
+		balloon_alert(user, "still printing")
+		return
```
Comment (MrMelbert) [inline code/modules/reagents/chemistry/machinery/chem_master.dm:207]:
```suggestion
		balloon_alert(user, "still printing!")
		return .
```

File: code/modules/reagents/chemistry/machinery/chem_master.dm [hunk 4]
```diff
@@ -114,39 +125,119 @@
...
-	return ITEM_INTERACT_BLOCKING
 
 /obj/machinery/chem_master/crowbar_act(mob/living/user, obj/item/tool)
+	if(user.combat_mode)
+		return NONE
+
+	. = ITEM_INTERACT_BLOCKING
+	if(is_printing)
+		balloon_alert(user, "still printing")
+		return
```
Comment (MrMelbert) [inline code/modules/reagents/chemistry/machinery/chem_master.dm:220]:
```suggestion
		balloon_alert(user, "still printing!")
		return .
```

File: code/modules/reagents/chemistry/machinery/chem_master.dm [hunk 5]
```diff
@@ -114,39 +125,119 @@
...
+ * * mob/living/user - the player trying to replace the beaker
+ * * obj/item/reagent_containers/new_beaker - the beaker we are trying to insert, swap with existing or remove if null
+ */
+/obj/machinery/chem_master/proc/replace_beaker(mob/living/user, obj/item/reagent_containers/new_beaker)
+	PRIVATE_PROC(TRUE)
 
-	return ..()
+	if(!QDELETED(beaker))
+		try_put_in_hand(beaker, user)
+		beaker = null
```
Comment (MrMelbert) [inline code/modules/reagents/chemistry/machinery/chem_master.dm:237]: This `beaker = null` should actually be redundant given `/Exited` implementation
```suggestion
		try_put_in_hand(beaker, user)
```

File: code/modules/reagents/chemistry/machinery/chem_master.dm [hunk 6]
```diff
@@ -194,270 +266,274 @@
 		ui.open()
 
 /obj/machinery/chem_master/ui_static_data(mob/user)
-	var/list/data = list()
-	data["categories"] = list()
+	. = list()
```
Comment (MrMelbert) [inline code/modules/reagents/chemistry/machinery/chem_master.dm:269]: Why are you changing this to the dot operator? I think I mentioned this in the last PR of your's but according to our style guide you should avoid using `.` like this unless it's for runtime safety. 
`var/list/data = list()` is fine.

File: code/modules/reagents/chemistry/machinery/chem_master.dm [hunk 7]
```diff
@@ -194,270 +266,274 @@
...
+	if(do_transfer)
+		if(target.is_reacting)
+			return FALSE
+		if(source.trans_to(target, amount, target_id = reagent))
+			. = TRUE
+	else if(source.remove_reagent(reagent, amount))
+		. = TRUE
+	if(.)
+		update_appearance(UPDATE_OVERLAYS)
+		use_power(active_power_usage)
```
Comment (MrMelbert) [inline code/modules/reagents/chemistry/machinery/chem_master.dm:415]:
```suggestion
	if(.)
		update_appearance(UPDATE_OVERLAYS)
		use_power(active_power_usage)
	return . 
```

**Comment by SyncIt21** [comment]
> why'd you delete `chemmaster_overlay_screen_analysis`?

Because I moved that to client side meaning the server or the machine no longer needs a specific icon state to show it's in analysis mode (because when in client side Player A could be Analyzing reagent information while Player B could be moving reagents so the overlay would inaccurately represent what 2 or more players could be doing at 1 movement).

It's no longer a mode that is specific to the machine but to the client now

File: code/modules/reagents/chemistry/machinery/chem_master.dm (hunk 6 above)
Comment (MrMelbert) [reply code/modules/reagents/chemistry/machinery/chem_master.dm:268]: Resolved without answer? orrr

Comment (SyncIt21) [reply code/modules/reagents/chemistry/machinery/chem_master.dm]: Somehow I missed this. My bad

**Comment by SyncIt21** [comment]
1 month and 78 rounds of testing!. This should have been merged by now.@MrMelbert please do the honors.

**Comment by Deadgebert** [comment]
this is a bit late, but only 13 patches can be printed at a time instead of the normal 50

**Comment by SyncIt21** [comment]
> this is a bit late, but only 13 patches can be printed at a time instead of the normal 50

That is intended with tier 1 parts. Tier 4 can print 50 patches. This is done to give upgrading micro lasers a purpose which previously did nothing. It increment in steps of 4 I.e. 13, 26, 39 and 50 for each tier

13 is high enough for most purposes anyway cause you can just print again immediately afterwards.

**Comment by Deadgebert** [comment]
> > this is a bit late, but only 13 patches can be printed at a time instead of the normal 50
> 
> That is intended with tier 1 parts. Tier 4 can print 50 patches. This is done to give upgrading micro lasers a purpose which previously did nothing. It increment in 4 steps i.e. 13, 26, 39 and 50 for each tier
> 
> 13 is high enough for most purposes anyway cause you can just print again immediately afterwards.

oh... okay :(

**Pull request merged**